// @sap/approuter/lib/backend-request/headers.js
'use strict';

var _ = require('lodash');
var url = require('url');
var headerUtil = require('../utils/header-util');
var pathUtil = require('../utils/path-util');

var HEADERS_TO_REMOVE = ['connection',
    'keep-alive',
    'public',
    'proxy-authenticate',
    'transfer-encoding',
    'upgrade',
    'sec-websocket-key',
    'host'];

exports.getHeaders = function (req, accessToken, destination) {
  var headers;
  var route;
  var tokens = {
    accessToken: accessToken
  };
  if (req) {
    headers = _.omit(req.headers, HEADERS_TO_REMOVE);
    tokens.accessToken = tokens.accessToken || (req.session && req.session.user && req.session.user.token &&
            req.session.user.token.accessToken);
    destination = destination || req.internalUrl.destination;

    addXForwardingHeaders(headers, req);
    removeSecurityHeaders(headers, req);
    headerUtil.updateSapPassport(headers);
    tokens['connProxyToken'] = req.app && req.app.connProxyToken && req.app.connProxyToken.accessToken;
    tokens['html5AppsRepoToken'] = req.app && req.app.html5AppsRepoToken && req.app.html5AppsRepoToken.accessToken;
    route = req.internalUrl ? req.internalUrl.route : null;
  }
  return addOauthHeader(headers || {}, tokens, route, destination);
};

function removeSecurityHeaders(headers, req) {
  if (!pathUtil.isPublicPath(req)) {
    delete headers['authorization'];
    if (pathUtil.isCsrfProtectionEnabled(req)) {
      delete headers['x-csrf-token'];
    }
  }
}

function addOauthHeader(headers, tokens, route, destination) {
  if (route && route.destination) { // Destination flow
    if (destination.proxyType && destination.proxyType === 'OnPremise') { // On premise destination flow - cloud connector
      headers['SAP-Connectivity-Authentication'] = 'Bearer ' + tokens.accessToken;
      headers['Proxy-Authorization'] = 'Bearer ' + tokens.connProxyToken;
      if (destination.cloudConnectorLocationId) {
        headers['SAP-Connectivity-SCC-Location_ID'] = destination.cloudConnectorLocationId;
      }
    } else { // Cloud destination flow
      if (destination.forwardAuthToken && destination.forwardAuthToken === true) {
        headers.authorization = 'Bearer ' + tokens.accessToken;
      }
    }
    // Common for on premise and cloud destination flows
    if (destination.authentication === 'BasicAuthentication' && destination.user && destination.password) {
      headers.authorization = 'Basic ' + Buffer(destination.user + ':' + destination.password).toString('base64');
    }
  } else if (route && route.service) {// Service flow
    if (route.service === 'html5AppsRepo') {// html5 apps repo flow
      headers['authorization'] = 'Bearer ' + tokens.html5AppsRepoToken;
    }
  } else {// WS or Logout
    if (tokens && tokens.accessToken) {
      headers.authorization = 'Bearer ' + tokens.accessToken;
    }
  }
  return headers;
}

function addXForwardingHeaders(headers, req) {
  if (req.headers.host) {
    headers['x-forwarded-host'] = req.headers.host;
  }
  if (!headers['x-forwarded-proto'] && req.protocol) {
    headers['x-forwarded-proto'] = req.protocol;
  }
  if (!headers['x-forwarded-for']) {
    headers['x-forwarded-for'] = req.connection.remoteAddress;
  }
  if (!headers['x-forwarded-path']) {
    headers['x-forwarded-path'] = url.parse(req.url).pathname;
  }
}
