// @sap/hdi/lib/main/ContainerGroupBase.js
'use strict';
const HDIBase = require('./HDIBase');
const CONSTANTS = require('../util/Constants');
const LibraryInformationResultTuple = require('../result/LibraryInformationResultTuple');
const typeCheck = require('../util/typeCheck');

const APIPrivilege = require('../parameter/APIPrivilege');
const LibraryConfiguration = require('../parameter/LibraryConfiguration');
const Parameter = require('../parameter/Parameter');
const SchemaPrivilege = require('../parameter/SchemaPrivilege');
const SchemaRole = require('../parameter/SchemaRole');


/**
 * Callback that takes errors as the first parameter and a APIPrivilege[] as the second parameter.
 * @callback APIPrivilegesCB
 * @param {Error} e Errors that ocurred. Is null if no errors occured.
 * @param {APIPrivilege[]} result The result. Is null if errors occured.
 */


/**
 * Base class for the {@link HDI} class and the {@link ContainerGroup} class.
 * @extends HDIBase
 */
class ContainerGroupBase extends HDIBase {
  /**
   * @param {Object} credentials an object containing the necessary information to connect to a HANA database
   * @param {string} credentials.host hostname for the connection
   * @param {number} credentials.port port for the connection
   * @param {string} credentials.user username for the connection
   * @param {string} credentials.password password of the user
   * @param {string[]} credentials.ca Certificates to use for the connection. Optional.
   * @param {string} paramSchema a schema that will be used to create temp tables to supply arguments to procedure calls
   * @param {string} apiSchema a schema that will be used for the api calls
   */
  constructor(credentials, paramSchema, apiSchema) {
    super(credentials, paramSchema, apiSchema);
  }

  /**
   * Creates a container with the given name.
   * Requires EXECUTE privilege on CREATE_CONTAINER
   * @param {string} name the container name
   * @param {Parameter[]} params parameters controlling the current procedure call
   * @param {ResultTupleCB} cb - The callback handling the result.
   * @returns {undefined} Returns nothing/undefined.
   * @async
   */
  createContainer(name, params, cb) {
    try {
      typeCheck(params, 'params', Parameter, cb);
    } catch (error) {
      return cb(error);
    }

    const command = 'CREATE_CONTAINER';
    const sql = `call ${this.apiSchema}.${command}('${name}', ${this.paramSchema}.${CONSTANTS.PARAMETERS_TABLE}, ?, ?, ?)`;
    this.fillHelperTable(CONSTANTS.PARAMETERS_TABLE, params)
      .then(() => this.executeCall(sql))
      .then((res) => cb(null, res))
      .catch((e) => cb(e));
  }

  /**
   * Drops the container with the given name, incl. all corresponding technical users, schemata, tables, etc.
   * Requires EXECUTE privilege on DROP_CONTAINER
   * @param {string} name the container name
   * @param {Parameter[]} params parameters controlling the current procedure call. Supported key-value combinations: IGNORE_WORK: true, false (default), IGNORE_DEPLOYED: true, false (default)
   * @param {ResultTupleCB} cb - The callback handling the result.
   * @returns {undefined} Returns nothing/undefined.
   * @async
   */
  dropContainer(name, params, cb) {
    try {
      typeCheck(params, 'params', Parameter);
    } catch (error) {
      return cb(error);
    }

    const command = 'DROP_CONTAINER';
    const sql = `call ${this.apiSchema}.${command}('${name}', ${this.paramSchema}.${CONSTANTS.PARAMETERS_TABLE}, ?, ?, ?)`;
    this.fillHelperTable(CONSTANTS.PARAMETERS_TABLE, params)
      .then(() => this.executeCall(sql))
      .then((res) => cb(null, res))
      .catch((e) => cb(e));
  }

  /**
   * Configures a container with the given configuration parameters and the given parameters.
   * Requires EXECUTE privilege on CONFIGURE_CONTAINER_PARAMETERS
   * @param {string} container The container name.
   * @param { Parameter[]} configParams parameters to configure the container
   * @param {Parameter[]} params parameters controlling the current procedure call
   * @param {ResultTupleCB} cb - The callback handling the result.
   * @returns {undefined} Returns nothing/undefined.
   * @async
   */
  configureContainerParameters(container, configParams, params, cb) {
    try {
      typeCheck(configParams, 'configParams', Parameter);
      typeCheck(params, 'params', Parameter);
    } catch (error) {
      return cb(error);
    }

    const command = 'CONFIGURE_CONTAINER_PARAMETERS';
    const sql = `call ${this.apiSchema}.${command}('${container}', ${this.paramSchema}.${CONSTANTS.CONFIG_PARAMETERS_TABLE}, ${this.paramSchema}.${CONSTANTS.PARAMETERS_TABLE}, ?, ?, ?)`;
    Promise.all([
      this.fillHelperTable(CONSTANTS.CONFIG_PARAMETERS_TABLE, configParams),
      this.fillHelperTable(CONSTANTS.PARAMETERS_TABLE, params)
    ])
      .then(() => this.executeCall(sql))
      .then((res) => cb(null, res))
      .catch((e) => cb(e));
  }

  /**
   * (Re)configures the set of plugin libraries which are installed in the given container.
   * Requires EXECUTE privilege on CONFIGURE_LIBRARIES
   * @param {string} container the affected container
   * @param {LibraryConfiguration[]} libs Configuration for libraries
   * @param {Parameter[]} params parameters controlling the current procedure call. Supported key-value combinations: UNDEPLOY: true, false (default)
   * @param {ResultTupleCB} cb - The callback handling the result.
   * @returns {undefined} Returns nothing/undefined.
   * @async
   */
  configureLibraries(container, libs, params, cb) {
    try {
      typeCheck(libs, 'libs', LibraryConfiguration);
      typeCheck(params, 'params', Parameter);
    } catch (error) {
      return cb(error);
    }

    const command = 'CONFIGURE_LIBRARIES';
    const sql = `call ${this.apiSchema}.${command}('${container}', ${this.paramSchema}.${CONSTANTS.LIBRARY_CONFIG_TABLE}, ${this.paramSchema}.${CONSTANTS.PARAMETERS_TABLE}, ?, ?, ?)`;
    Promise.all([
      this.fillHelperTable(CONSTANTS.LIBRARY_CONFIG_TABLE, libs),
      this.fillHelperTable(CONSTANTS.PARAMETERS_TABLE, params)
    ])
      .then(() => this.executeCall(sql))
      .then((res) => cb(null, res))
      .catch((e) => cb(e));
  }

  /**
   * List the set of plugin libraries which are installed in the given container.
   * Requires EXECUTE privilege on LIST_CONFIGURED_LIBRARIES
   * @param {string} container the affected container.
   * @param {Parameter[]} params parameters controlling the current procedure call
   * @param {ResultTupleCB} cb - The callback handling the result.
   * @returns {undefined} Returns nothing/undefined.
   * @async
   */
  listConfiguredLibraries(container, params, cb) {
    try {
      typeCheck(params, 'params', Parameter);
    } catch (error) {
      return cb(error);
    }

    const command = 'LIST_CONFIGURED_LIBRARIES';
    const sql = `call ${this.apiSchema}.${command}('${container}', ${this.paramSchema}.${CONSTANTS.PARAMETERS_TABLE}, ?, ?, ?, ?)`;
    const result = new LibraryInformationResultTuple();
    this.fillHelperTable(CONSTANTS.PARAMETERS_TABLE, params)
      .then(() => this.executeCall(sql, result))
      .then((res) => cb(null, res))
      .catch((e) => cb(e));
  }

  /**
   * Tries to cancel the asynchronous request with the given request id.
   * Requires EXECUTE privilege on CANCEL
   * @param {string} container the affected container.
   * @param {number} requestId request id of the request to cancel
   * @param {Parameter[]} params parameters controlling the current procedure call
   * @param {ResultTupleCB} cb - The callback handling the result.
   * @returns {undefined} Returns nothing/undefined.
   * @async
   */
  cancel(container, requestId, params, cb) {
    try {
      typeCheck(params, 'params', Parameter);
    } catch (error) {
      return cb(error);
    }

    const command = 'CANCEL';
    const sql = `call ${this.apiSchema}.${command}('${container}', ${requestId}, ${this.paramSchema}.${CONSTANTS.PARAMETERS_TABLE}, ?, ?, ?)`;
    this.fillHelperTable(CONSTANTS.PARAMETERS_TABLE, params)
      .then(() => this.executeCall(sql))
      .then((res) => cb(null, res))
      .catch((e) => cb(e));
  }

  /**
   * Grants the given privileges on the container's API objects to the given users.
   * Requires EXECUTE privilege on GRANT_CONTAINER_API_PRIVILEGES
   * @param {string} container the affected container.
   * @param {APIPrivilege[]} privileges list of APIPrivilege describing what and to whom to grant
   * @param {Parameter[]} params parameters controlling the current procedure call
   * @param {ResultTupleCB} cb - The callback handling the result.
   * @returns {undefined} Returns nothing/undefined.
   * @async
   */
  grantContainerApiPrivileges(container, privileges, params, cb) {
    try {
      typeCheck(privileges, 'privileges', APIPrivilege);
      typeCheck(params, 'params', Parameter);
    } catch (error) {
      return cb(error);
    }
    const command = 'GRANT_CONTAINER_API_PRIVILEGES';
    const sql = `call ${this.apiSchema}.${command}('${container}', ${this.paramSchema}.${CONSTANTS.API_PRIVILEGES_TABLE}, ${this.paramSchema}.${CONSTANTS.PARAMETERS_TABLE}, ?, ?, ?)`;
    Promise.all([
      this.fillHelperTable(CONSTANTS.API_PRIVILEGES_TABLE, privileges),
      this.fillHelperTable(CONSTANTS.PARAMETERS_TABLE, params)
    ])
      .then(() => this.executeCall(sql))
      .then((res) => cb(null, res))
      .catch((e) => cb(e));
  }

  /**
   * Grants the given privileges on the container's API objects to the given users with grant option.
   * Requires EXECUTE privilege on GRANT_CONTAINER_API_PRIVILEGES_WITH_GRANT_OPTION
   * @param {string} container the affected container.
   * @param {APIPrivilege[]} privileges list of APIPrivilege describing what and to whom to grant
   * @param {Parameter[]} params parameters controlling the current procedure call
   * @param {ResultTupleCB} cb - The callback handling the result.
   * @returns {undefined} Returns nothing/undefined.
   * @async
   */
  grantContainerApiPrivilegesWithGrantOption(container, privileges, params, cb) {
    try {
      typeCheck(privileges, 'privileges', APIPrivilege);
      typeCheck(params, 'params', Parameter);
    } catch (error) {
      return cb(error);
    }

    const command = 'GRANT_CONTAINER_API_PRIVILEGES_WITH_GRANT_OPTION';
    const sql = `call ${this.apiSchema}.${command}('${container}', ${this.paramSchema}.${CONSTANTS.API_PRIVILEGES_TABLE}, ${this.paramSchema}.${CONSTANTS.PARAMETERS_TABLE}, ?, ?, ?)`;
    Promise.all([
      this.fillHelperTable(CONSTANTS.API_PRIVILEGES_TABLE, privileges),
      this.fillHelperTable(CONSTANTS.PARAMETERS_TABLE, params)
    ])
      .then(() => this.executeCall(sql))
      .then((res) => cb(null, res))
      .catch((e) => cb(e));
  }

  /**
   * Revokes the given privileges on the container's API objects from the given users.
   * Requires EXECUTE privilege on REVOKE_CONTAINER_API_PRIVILEGES
   * @param {string} container the affected container.
   * @param {APIPrivilege[]} privileges list of APIPrivilege
   * @param {Parameter[]} params parameters controlling the current procedure call
   * @param {ResultTupleCB} cb - The callback handling the result.
   * @returns {undefined} Returns nothing/undefined.
   * @async
   */
  revokeContainerApiPrivileges(container, privileges, params, cb) {
    try {
      typeCheck(privileges, 'privileges', APIPrivilege);
      typeCheck(params, 'params', Parameter);
    } catch (error) {
      return cb(error);
    }

    const command = 'REVOKE_CONTAINER_API_PRIVILEGES';
    const sql = `call ${this.apiSchema}.${command}('${container}', ${this.paramSchema}.${CONSTANTS.API_PRIVILEGES_TABLE}, ${this.paramSchema}.${CONSTANTS.PARAMETERS_TABLE}, ?, ?, ?)`;
    Promise.all([
      this.fillHelperTable(CONSTANTS.API_PRIVILEGES_TABLE, privileges),
      this.fillHelperTable(CONSTANTS.PARAMETERS_TABLE, params)
    ])
      .then(() => this.executeCall(sql))
      .then((res) => cb(null, res))
      .catch((e) => cb(e));
  }

  /**
   * Grants the given privileges on the container's target schema to the given users.
   * Requires EXECUTE privilege on GRANT_CONTAINER_SCHEMA_PRIVILEGES
   * @param {string} container the affected container.
   * @param {SchemaPrivilege[]} privileges list of SchemaPrivilege describing what and to whom to grant
   * @param {Parameter[]} params parameters controlling the current procedure call
   * @param {ResultTupleCB} cb - The callback handling the result.
   * @returns {undefined} Returns nothing/undefined.
   * @async
   */
  grantContainerSchemaPrivileges(container, privileges, params, cb) {
    try {
      typeCheck(privileges, 'privileges', SchemaPrivilege);
      typeCheck(params, 'params', Parameter);
    } catch (error) {
      return cb(error);
    }

    const command = 'GRANT_CONTAINER_SCHEMA_PRIVILEGES';
    const sql = `call ${this.apiSchema}.${command}('${container}', ${this.paramSchema}.${CONSTANTS.SCHEMA_PRIVILEGES_TABLE}, ${this.paramSchema}.${CONSTANTS.PARAMETERS_TABLE}, ?, ?, ?)`;
    Promise.all([
      this.fillHelperTable(CONSTANTS.SCHEMA_PRIVILEGES_TABLE, privileges),
      this.fillHelperTable(CONSTANTS.PARAMETERS_TABLE, params)
    ])
      .then(() => this.executeCall(sql))
      .then((res) => cb(null, res))
      .catch((e) => cb(e));
  }

  /**
   * Revokes the given privileges on the container's target schema from the given users.
   * Requires EXECUTE privilege on REVOKE_CONTAINER_SCHEMA_PRIVILEGES
   * @param {string} container the affected container.
   * @param {SchemaPrivilege} privileges list of SchemaPrivilege
   * @param {Parameter[]} params parameters controlling the current procedure call
   * @param {ResultTupleCB} cb - The callback handling the result.
   * @returns {undefined} Returns nothing/undefined.
   * @async
   */
  revokeContainerSchemaPrivileges(container, privileges, params, cb) {
    try {
      typeCheck(privileges, 'privileges', SchemaPrivilege);
      typeCheck(params, 'params', Parameter);
    } catch (error) {
      return cb(error);
    }
    const command = 'REVOKE_CONTAINER_SCHEMA_PRIVILEGES';
    const sql = `call ${this.apiSchema}.${command}('${container}', ${this.paramSchema}.${CONSTANTS.SCHEMA_PRIVILEGES_TABLE}, ${this.paramSchema}.${CONSTANTS.PARAMETERS_TABLE}, ?, ?, ?)`;
    Promise.all([
      this.fillHelperTable(CONSTANTS.SCHEMA_PRIVILEGES_TABLE, privileges),
      this.fillHelperTable(CONSTANTS.PARAMETERS_TABLE, params)
    ])
      .then(() => this.executeCall(sql))
      .then((res) => cb(null, res))
      .catch((e) => cb(e));
  }

  /**
   * Grants the given roles which are deployed inside the container to the given users.
   * Requires EXECUTE privilege on GRANT_CONTAINER_SCHEMA_ROLES
   * @param {string} container the affected container.
   * @param {SchemaRole[]} roles list of SchemaRole
   * @param {Parameter[]} params parameters controlling the current procedure call
   * @param {ResultTupleCB} cb - The callback handling the result.
   * @returns {undefined} Returns nothing/undefined.
   * @async
   */
  grantContainerSchemaRoles(container, roles, params, cb) {
    try {
      typeCheck(roles, 'roles', SchemaRole);
      typeCheck(params, 'params', Parameter);
    } catch (error) {
      return cb(error);
    }
    const command = 'GRANT_CONTAINER_SCHEMA_ROLES';
    const sql = `call ${this.apiSchema}.${command}('${container}', ${this.paramSchema}.${CONSTANTS.SCHEMA_ROLES_TABLE}, ${this.paramSchema}.${CONSTANTS.PARAMETERS_TABLE}, ?, ?, ?)`;
    Promise.all([
      this.fillHelperTable(CONSTANTS.SCHEMA_ROLES_TABLE, roles),
      this.fillHelperTable(CONSTANTS.PARAMETERS_TABLE, params)
    ])
      .then(() => this.executeCall(sql))
      .then((res) => cb(null, res))
      .catch((e) => cb(e));
  }

  /**
   * Revokes the given roles which are deployed inside the container from the given users.
   * Requires EXECUTE privilege on REVOKE_CONTAINER_SCHEMA_ROLES
   * @param {string} container the affected container.
   * @param {SchemaRole[]} roles list of SchemaRole
   * @param {Parameter[]} params parameters controlling the current procedure call
   * @param {ResultTupleCB} cb - The callback handling the result.
   * @returns {undefined} Returns nothing/undefined.
   * @async
   */
  revokeContainerSchemaRoles(container, roles, params, cb) {
    try {
      typeCheck(roles, 'roles', SchemaRole);
      typeCheck(params, 'params', Parameter);
    } catch (error) {
      return cb(error);
    }
    const command = 'REVOKE_CONTAINER_SCHEMA_ROLES';
    const sql = `call ${this.apiSchema}.${command}('${container}', ${this.paramSchema}.${CONSTANTS.SCHEMA_ROLES_TABLE}, ${this.paramSchema}.${CONSTANTS.PARAMETERS_TABLE}, ?, ?, ?)`;
    Promise.all([
      this.fillHelperTable(CONSTANTS.SCHEMA_ROLES_TABLE, roles),
      this.fillHelperTable(CONSTANTS.PARAMETERS_TABLE, params)
    ])
      .then(() => this.executeCall(sql))
      .then((res) => cb(null, res))
      .catch((e) => cb(e));
  }

  /**
   * Returns the default privileges required for administration of a container group
   * Requires SELECT privilege on _SYS_DI.T_DEFAULT_CONTAINER_GROUP_ADMIN_PRIVILEGES
   * Minimum required HANA server version: 2.00.010
   * @param {string} principalSchema schema of the target user/role (empty for global principals)
   * @param {string} principalName name of the target user/role
   * @param {APIPrivilegesCB} cb - The callback handling the result.
   * @returns {undefined} Returns nothing/undefined.
   * @async
   */
  getDefaultContainerGroupAdminPrivileges(principalSchema, principalName, cb) {
    const sql = `select PRIVILEGE_NAME, OBJECT_NAME from ${CONSTANTS.GROUP_ADMIN_PRIVILEGES_TABLE}`;
    this.getAPIPrivileges(sql, principalSchema, principalName)
      .then((res) => cb(null, res))
      .catch((e) => cb(e));
  }
}
module.exports = ContainerGroupBase;
